### PR TITLE
[WIP] Incorporate librdkafka into cudf CMake

### LINKED ISSUE #2473: [QST] - Propose directly invoking librdkafka from libcudf
**What is your question?**

kafka_current
kafka_proposed

There is currently a substantial bottleneck when reading messages from Kafka -> librdkafka (C kafka library) -> python-confluent-kafka (python kafka library that wraps librdkafka) -> Python -> and then to cuDF. 

In a nutshell the slowest part is the creation of the PyObjects that are returned from the librdkafka C library back to Python. In the case of cuDF this bottleneck could be completely avoided as we simply receive those PyObjects and turn around to immediately place them on the GPU as a cuDF DataFrame.

I wanted to discuss the possibility of simply overloading the cudf.from_csv(), from_json(), etc functions to allow for libcudf to read directly from librdkafka and circumvent the Python library completely. This would greatly speed up our streaming strategy and allow for larger datasets to be placed directly on the GPU without the creation of PyObjects.

I have attached 2 rough whiteboard screenshots (sorry for the terrible light glare). Curious what others thoughts are on this? The librdkafka library is quite robust and sound be simple to integrate into libcudf.

### PR DESCRIPTION
**Pull request by jdye64**
There is no code in this PR but rather just to validate with the larger group that this is indeed the proper way we want to do this

Closes #2473 

@harrism This is just a WIP PR that simply adds the librdkafka library to the current cudf build. I was hoping you could validate what I am doing here is ok. IF it is then I will continue on with the implementation details.

### CODE REVIEW

**Comment by harrism** [comment]
Moving to 0.10 since this is WP and 0.9 is already in burn down.

**Comment by harrism** [comment]
the cmake looks reasonable. @kkraus14 agree?

**Comment by jdye64** [comment]
I will add that I did not link zlib or lz4 yet which would probably be best to also add but just wanted a baseline first.

**Comment by kkraus14** [comment]
Is there a reason we're building librdkafka from source instead of just depending on an install via conda or other means? I would rather treat it similarly to how we treat ZLIB in the cmake for example and then depend on https://anaconda.org/conda-forge/librdkafka in the conda environments / recipes.

**Comment by jdye64** [comment]
@kkraus14 I debated both. I don't really care either way but here is why I was thinking cmake build from source. Its really not that many differences.

1) Its a small and quick build so the added build time wasn't minimal.
2) Allows for us to use a very specific version vs what is installed via a package manager hence allowing us to control which version we pull from Git and use between releases OR as needed to test new features without relying on the anaconda packages since those are out of our control. I see lots of other projects struggle with this given the rapid change in between Kafka releases and version requirements.
3) Saved the end user from performing another conda install therefore less docs and a more "simple" installation process.

I'm fine with either way just wanted to point out why I chose this route.

**Comment by kkraus14** [comment]
> @kkraus14 I debated both. I don't really care either way but here is why I was thinking cmake build from source. Its really not that many differences.
> 
> 1. Its a small and quick build so the added build time wasn't minimal.
> 2. Allows for us to use a very specific version vs what is installed via a package manager hence allowing us to control which version we pull from Git and use between releases OR as needed to test new features without relying on the anaconda packages since those are out of our control. I see lots of other projects struggle with this given the rapid change in between Kafka releases and version requirements.
> 3. Saved the end user from performing another conda install therefore less docs and a more "simple" installation process.
> 
> I'm fine with either way just wanted to point out why I chose this route.

But what if a user installs confluent-kafka-python for usage elsewhere, which would then install a separate librdkafka with a potentially different version and then it becomes a race as to which symbols are loaded first, no?

**Comment by jdye64** [comment]
I opted to use the CMake ```NO_DEFAULT_PATH``` option which should prevent that when locating the librdkafka library and headers.

https://github.com/rapidsai/cudf/pull/2519/commits/cb4228afa6bbaac2d3b2efdbe9e6abc5f5e227bd#diff-e06d4dfe6ac984f1aed235fa6161fcc3R49

File: .gitignore [hunk 1]
```diff
@@ -66,6 +66,7 @@ build/
 cpp/build/
 cpp/include/cudf/ipc_generated/*.h
 cpp/thirdparty/googletest/
+cpp/cmake-build-debug
```
Comment (harrism) [inline .gitignore]: This folder is specific to you.  Please use cpp/Debug, which is already .gitignored.

File: cpp/src/io/io_functions.cpp [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #include <cudf/cudf.h>
 #include <cudf/legacy/table.hpp>
 #include <utilities/error_utils.hpp>
+#include <librdkafka/rdkafkacpp.h>
```
Comment (harrism) [inline cpp/src/io/io_functions.cpp]: Unused include file?

```suggestion
```

**Comment by harrism** [comment]
I wouldn't want to merge this until the functionality is actually used. Otherwise it's an unnecessary dependency.  So I would extend this PR to actually implement the functionality.

**Comment by harrism** [comment]
As for where the logic should go, you might want to start a conversation with @mjsamoht @j-ieong @OlivierNV @vuule.

**Comment by OlivierNV** [comment]
Extending datasource indeed seems like the right place to put this.

**Comment by harrism** [comment]
@jdye64 can you drive this to completion in 0.11?

**Comment by jdye64** [comment]
I'm going to close this as it is taken care of in https://github.com/rapidsai/cudf/pull/3504 now. Thanks for all the feedback here everyone!

**Pull request closed without merging**